### DISCUSSION TITLE
How to correctly validate a signature with a timestamp signed before revoking of the signature certificate?

### DISCUSSION THREAD

**Discussion by Mato-Z** [MatthiasValvekens/pyHanko#228] [Q&A]
Hello, I have question about validate a signature with a timestamp signed before later revocation of signature certificate (the X.509 file). For example I have PDF files (PKCS#7_T or PAdES_BASELINE_T) which was signed with PKCS#12 file (X.509 file with .pfx or .p12 extension) and during signing timestamps were correctly added. After few weeks X.509 certificate used for signing was revoked. During validation pyhanko evaluates the signature as invalid because signer certificate was revocated. But revocation time is older like signing time in timestamp. 
One specific example: This document was signed 2022-12-15T13:41:05+00:00 and cert was revoked revoked at 17:37:53 on 2023-02-08, so in time of signing cert was valid. Why is the signature evaluate like invalid please? 
It's possible that I just misunderstood the documentation. Is it possible, please, to perform validation so that the signature is evaluated as valid?
Many thanks. 
```
2023-02-24 20:11:03,559 - pyhanko.sign.validation.generic_cms - WARNING - OCSP response indicates the end-entity certificate was revoked at 17:37:53 on 2023-02-08, due to an unspecified reason.
2023-02-24 20:11:03,559 - pyhanko.sign.validation.generic_cms - WARNING - Chain of trust validation for ... failed.
===================
Field 1: Signature2
===================


Signer info
-----------
...
Trust anchor: "No path to trust anchor found."
The signer's certificate is revoked.


Integrity
---------
The signature is cryptographically sound.

The digest algorithm used was 'sha256'.
The signature mechanism used was 'sha256_rsa'.


Signing time
------------
Signing time as reported by signer: 2022-12-15T14:41:19+01:00

Signature timestamp token: 2022-12-15T13:41:05+00:00
The token is guaranteed to be newer than the signature.
This timestamp is backed by a time stamping authority.
The timestamp token is cryptographically sound.
...
The TSA certificate is trusted.

Content timestamp token: 2022-12-15T13:41:05+00:00
The token is guaranteed to be older than the signature.
This timestamp is backed by a time stamping authority.
The timestamp token is cryptographically sound.
...
The TSA certificate is trusted.


Modifications
-------------
The signature does not cover the entire file.
All modifications relate to signature maintenance, and they appear to be compatible with the current document modification policy.


Bottom line
-----------
The signature is judged INVALID.


Error: Validation failed
```

**Comment by MatthiasValvekens** [comment]
Hmm, this definitely _should_ work if the signature was generated correctly---there are unit tests for exactly this kind of scenario... I'm not sure why it fails. Did you embed revocation information at the time of signing? Can you try validating against an older version of pyHanko and `pyhanko-certvalidator`?

------------------------

Having said that, the LTV validator currently exposed in the CLI is [not very good](https://pyhanko.readthedocs.io/en/0.17.0/known-issues.html). It has numerous issues, and it's very particular about where to source revocation information from for basically no good reason.

The latest version includes an experimental AdES validation engine that is definitely _more_ correct, but (a) that one is still incubating and more difficult to use without specialised knowledge, and (b) not exposed in the CLI at all. If you're up for it, feel free to give it a try. The plan is to deprecate the old LTV validator as soon as the AdES validator is stable. Not sure when that'll happen, though.

**Reply by Mato-Z** [reply]
@MatthiasValvekens thanks for your quick reply. With older version of pyhanko - 0.16.0 result is the same. 
Revocation information did not embed in the time of signing. Is this a problem? These are not my documents, they were created in Adobe Acrobat and LibreOffice. I just need to validate them correctly. I can send them to you internally via email for possible deeper investigation.
Many thanks.

**Reply by MatthiasValvekens** [reply]
> Revocation information did not embed in the time of signing. Is this a problem?

Whether this is a problem or not depends on the strictness of your validation policy, but it's true that it's generally expected.

On reflection, I don't think it's the cause for the behaviour you're seeing, though. More likely than not, the issue is far more basic: the validation logic called by the CLI will not even bother to try validating the signature at the time in the timestamp. Can you try again with `--ltv-profile pades` or `--ltv-profile adobe`?

(note: even if that works, it's not a perfect solution, because there will be cases where the LTV validator fails but the "plain" one succeeds. Like I said, it's not very good.)

Either way, the situation should improve as the AdES validation engine matures, but currently I'm afraid pyHanko doesn't really have a "plug-and-play" general-purpose validator that will magically do the right thing in all of these cases... :/

There's an old TODO comment in the code somewhere to implement a validator that would autodetect the right signature profile and apply that, but I never acted on that because the underlying issues with the validation code would've gotten in the way anyhow.

**Reply by Mato-Z** [reply]
@MatthiasValvekens Thanks. Interesting if I use `--ltv-profile pades` on PAdES_BASELINE_T or PKCS#7 format I get this output:
```
2023-02-25 18:11:18,736 - pyhanko.cli - ERROR - An error occurred while parsing the revocation information for this signature: 
===================
Field 1: Signature2
===================


An error occurred while parsing the revocation information for this signature: 
Error: Validation failed
```

And If I use `--ltv-profile adobe` on PAdES_BASELINE_T or PKCS#7 format I get this output:
```
2023-02-25 18:10:47,276 - pyhanko.sign.validation.generic_cms - WARNING - The path could not be validated because no revocation information could be found for intermediate certificate 1
2023-02-25 18:10:47,276 - pyhanko.sign.validation.generic_cms - WARNING - Chain of trust validation for Common Name: ... failed.
2023-02-25 18:10:47,281 - pyhanko.sign.validation.generic_cms - WARNING - The path could not be validated because no revocation information could be found for intermediate certificate 1
2023-02-25 18:10:47,281 - pyhanko.sign.validation.generic_cms - WARNING - Chain of trust validation for Email Address: ... failed.

```
I need to validate signed theses from students using X.509 certificates issued by an authority that doesn't use them directly for signing documents, but for S/MIME. However, it is only for our internal needs. The timestamp is also from this time authority. In the signing certificates, there is an intermediate entity and root authority USERTrust RSA Certification Authority. With a non-revoked certificate, everything is ok, I had to modify the config as follows:
```
validation-contexts:
     setup-xy:
         signer-key-usage: ["digital_signature"]
         trust: TSA_Authority_Root.pem
         time-tolerance: 180
```
Should I modify the config somehow?
I also sent you sample of signed files via email.
Many thanks.

**Reply by MatthiasValvekens** [reply]
Hi @Mato-Z, I haven't forgotten, but it might be a few more days before I can look into this properly. Thanks for sending the files; that'll definitely help.

**Reply by Mato-Z** [reply]
Hi @MatthiasValvekens it is OK, many thanks for your help.

**Reply by Mato-Z** [reply]
Hello @MatthiasValvekens any progress please?

**Reply by MatthiasValvekens** [reply]
Unfortunately not yet, I've been quite busy. Your email is still flagged in my inbox, though, so I'll get to it eventually :).

**Reply by Mato-Z** [reply]
@MatthiasValvekens ok, thanks for info :-)

**Reply by MatthiasValvekens** [reply]
Hi @Mato-Z,

Quick update: the bad news is that the "old" validation engine exposed in the CLI is too rigid to validate your signature: it doesn't deal very well with cases where part of the validation info comes from within the file, and the rest needs to be fetched live. The good news is that it looks like the new experimental AdES LTA validator is capable of handling it, but I found a few (hopefully minor) issues that need to be addressed there before I can make a definite judgment.

EDIT: I have to run now, but I'll try to make some time tonight to investigate those remaining issues.

**Reply by Mato-Z** [reply]
Hi @MatthiasValvekens Thanks a lot for your information and help. So if minor problems will be solved (what I hope for)  I can try validation also using the API (SDK) without CLI.

**Reply by MatthiasValvekens** [reply]
Hi @Mato-Z,

I think I fixed the remaining bugs I mentioned earlier. They were all ultimately straightforward data flow/presentation issues, fortunately (so the validation engine itself still did the right thing), but I ended up having to expose some additional things in `pyhanko-certvalidator`, so it went a little deeper than anticipated.

With the fixes (see [this branch](https://github.com/MatthiasValvekens/pyHanko/tree/bugfix/ades-revoked-cert-validation-issues)), the following snippet does the job:


```python
import asyncio

from pyhanko.keys import load_cert_from_pemder
from pyhanko.pdf_utils.reader import PdfFileReader
from pyhanko.sign.validation import ades
from pyhanko.sign.validation.policy_decl import SignatureValidationSpec, PdfSignatureValidationSpec
from pyhanko_certvalidator.context import CertValidationPolicySpec
from pyhanko_certvalidator.policy_decl import FreshnessReqType
from pyhanko_certvalidator.registry import SimpleTrustManager
from pyhanko_certvalidator import policy_decl as certv_policy_decl

TSA_ROOT_CERT_FILE = '/tmp/pyhk228/pyhk228-tsa-root.crt'
TEST_PDF = '/tmp/pyhk228/pyhk228-pades.pdf'

tsa_root = load_cert_from_pemder(TSA_ROOT_CERT_FILE)
the_policy = PdfSignatureValidationSpec(
    signature_validation_spec=SignatureValidationSpec(
        cert_validation_policy=CertValidationPolicySpec(
            trust_manager=SimpleTrustManager.build(extra_trust_roots=[tsa_root]),
            revinfo_policy=certv_policy_decl.CertRevTrustPolicy(
                revocation_checking_policy=certv_policy_decl.REQUIRE_REVINFO,
                freshness_req_type=FreshnessReqType.MAX_DIFF_REVOCATION_VALIDATION,
            ),
        )
    ),
)


async def validate():
    with open(TEST_PDF, 'rb') as inf:
        r = PdfFileReader(inf)
        result = await ades.ades_lta_validation(
            r.embedded_signatures[0],
            pdf_validation_spec=the_policy,
        )
        print(result.api_status.pretty_print_details())
        print(f"AdES result: {result.ades_subindic}")
        print(f"Best signature time: {result.best_signature_time}")


if __name__ == '__main__':
    asyncio.run(validate())
```

When run, this produces:

```
No document timestamps after signature; proceeding without past proof of existence
OCSP response indicates the end-entity certificate was revoked at 16:30:14 on 2023-02-08, due to certificate supersession.

<info redacted for privacy reasons>

Modifications
-------------
The signature covers the entire file.


Bottom line
-----------
The signature is judged VALID.


AdES result: AdESPassed.OK
Best signature time: 2023-02-08 16:24:57+00:00

```

Without the TSA root, you end up with this:

```
<a bunch of warnings about PoE handling and timestamp chains>
<snip>

Bottom line
-----------
The signature is judged INVALID.


AdES result: AdESIndeterminate.NO_POE
Best signature time: <current time>
```

Still need to write some regression tests, but I believe we're on the right track :).

**Reply by MatthiasValvekens** [reply]
See #251 .

**Reply by MatthiasValvekens** [reply]
I just merged #251, so you should be able to do what you want with the new AdES validator starting from the next release :).

**Reply by Mato-Z** [reply]
@MatthiasValvekens  Great! Many many thanks! I cloned your repository and tried it without CLI. It works perfectly :-) Now I will have to test other things, mainly to automate the addition of additional time stamps for long-term archiving (PAdES BASELINE-LTA). We need to preserve works for a long time so that their credibility can be proven. I also need to test the addition of validation information ex-post (PAdES BASELINE-LT), because Adobe Acrobat is not able to add it during the signature itself with the certificates issued by our authority (you saw it), and unfortunately many students cannot use any other tool. I hope I can do it all with pyHanko. Then I can add a practical usage scenario :-)

**Reply by MatthiasValvekens** [reply]
Sounds good! This is one of the core use cases that the B-LTA profile of PAdES was designed to handle, so you should be able to make it work.

The only tricky aspect that I can think of right now is that you possibly won't be able to validate the full chain at the current time when you receive the document (given that the leaf certificates are very short-lived). Nonetheless, the method used in [add_validation_info](https://github.com/MatthiasValvekens/pyHanko/blob/master/pyhanko/sign/validation/__init__.py#L276) applies more or less (mutatis mutandis---you might have to re-roll part of the revinfo fetching logic to accommodate your PKI's idiosyncrasies).

Thinking out loud: there might still be issues if (unlike in the example you provided me) the revocation info for the leaf certs is no longer available and the signer hasn't provided any. AdES has some provisions for post-expiration revocation data, but it's tricky stuff---feel free to make another post if you have a practical use case for that ;). 

To add the document timestamp signature required for LTA, you can just use `PdfTimeStamper`.

**Reply by Mato-Z** [reply]
@MatthiasValvekens thanks. Of course I will definitely inform you about progress in short time.

**Reply by Mato-Z** [reply]
@MatthiasValvekens I have a problem if I want to add validation information when signing, and I use the following procedure: 
`pyhanko --verbose --config pyhanko.yml sign addsig --field Sig1 --timestamp-url MY_TSA --with-validation-info --validation-context setup-b --use-pades pkcs12 test.pdf output.pdf MY_PKCS12_FILE `
The output file is only in PAdES BASELINE-T format and not PAdES BASELINE-LT as you describe https://pyhanko.readthedocs.io/en/latest/cli-guide/signing.html#embedding-revocation-info-with-pyhanko. I sent the file created in this way by e-mail with a detailed log from the generation.
And one more small question: Is this sentence in the validation output correct? `No document timestamps after signature; proceeding without past proof of existence`, when timestamp was automatically added during the signing? Isn't it a bit confusing even if it doesn't affect the actual verification result? It is a common behaviour of various tools like pyhanko, or Adobe Acrobat, etc. when signing (PAdES BASELINE-T)...

**Reply by Mato-Z** [reply]
@MatthiasValvekens When I try: `pyhanko --verbose --config pyhanko.yml sign ltvfix --field Sig1 --timestamp-url MY_TSA --validation-context setup-b output.pdf` format is still only BASELINE-T.

**Reply by MatthiasValvekens** [reply]
Hi @Mato-Z,

Sorry for the delay in replying. The situation you are describing arises because some revocation info in the PDF is lacking. As the CLI guide explains, `ltvfix` doesn't actually _guarantee_ PADES-B-LT, it merely makes a good faith attempt at doing so.

That being said, I have a potential explanation/solution in this case: last week I found a regression in `pyhanko-certvalidator` in the sanity checks on incoming revocation data. The way that usually manifests is that pyHanko rejects otherwise valid OCSP responses, which could cause the behaviour you're seeing. The bug in question should be fixed in `pyhanko-certvalidator` `0.21.x` and up. Can you try upgrading `pyhanko-certvalidator` to `0.22.0`? You might get an error from `pip` about dependency mismatches, but it's mostly safe to ignore---the breaking change is a very minor one that doesn't really matter for CLI users.

PyHanko `0.18.0` will be released very soon, and it includes a bump for `pyhanko-certvalidator`, so if you can wait a few more days, that should also do the trick.

If you still see the same issue after bumping `pyhanko-certvalidator`, let me know.

**Reply by Mato-Z** [reply]
@MatthiasValvekens many thanks for reply. But unfortunately  with pyhanko-certvalidator Version: 0.22.0 and pyHanko
Version: 0.18.1problem is the same :-( 
Again I sent email with detailed information and files... Thanks a lot for your help.

**Reply by Mato-Z** [reply]
@MatthiasValvekens with pyhanko version 0.21.0 adding validation/revocation information is ok and PDF file has correct format PAdES B-LT.
